File: Java/SieveOfEratosthenes.java
import java.util.*;
public class SieveOfEratosthenes{
    
    public static void main(String[] arg){

        Scanner sc= new Scanner(System.in);
        System.out.println("Enter n");
        int n = sc.nextInt();
        //Assume all No. 0 to n is prime(true)
        boolean prime[] = new boolean[n+1];
        Arrays.fill(prime, true);
        prime[0] = prime[1] = false;
        //check whether a no. is prime or not 
        for(int i=2;i*i<=n;i++){
            if(prime[i]){
                // if no. is not prime then the all of it's multiples 
                //are not prime so make them not prime(false)
                for(int j = i*i;j<=n;j+=i){
                    prime[j] = false;
                }
            }
        }
        //Print all prime numbers less then or equal to n
        for(int i=2;i<=n;i++){
            if(prime[i]){
                System.out.print(i + " ");
            }
        }
        System.out.println();
    }
}

//Given a number n, print all primes smaller than or equal to n using Sieve of Eratosthenes technique.
//Time complexity O(nlogn)
//space compexity O(n)

File: C/Sieve_of_Eratosthenes.c
#include <stdio.h>
#include <stdlib.h>

void seive_of_era(int n){
    long long i,j;
    int *primes;
    primes = malloc(sizeof(int) * (n+1));
    for (i = 2;i<n+1; i++){
        primes[i] = 1;
    }

    for (i = 2;i <n+1; i++){
        if (primes[i]){
            for (j=i;i*j<n+1; j++){
                primes[i * j] = 0;
            }
        }
    }
    printf("Prime number less than or equal to are: \n");
    for (i=2;i<n+1;i++){
        if (primes[i]){
            printf("%d ", i);
        }
    }
}
int main(void) {
	int n;
	printf("Enter the Number: ");
	scanf("%d",&n);
	seive_of_era(n);
	return 0;
}

/* 
The Sieve of Eratosthenes algorithm takes O(N log (log N)) time
to find all prime numbers less than N time complexity.
Space Complexity: O(1)

 Input and Output Cases:

 Enter the Number: 20
 Prime number less than or equal to are:
 2 3 5 7 11 13 17 19

 Enter the Number: 30
 Prime number less than or equal to are:
 2 3 5 7 11 13 17 19 23 29

 */
